### ISSUE TITLE
🐞 Container Gap values are not transferred when Cross-site Copying and Pasting 

### ISSUE THREAD

**Issue by KingGrizzly** [elementor/elementor#24111]
### Prerequisites

- [X] I have searched for similar issues in both open and closed tickets and cannot find a duplicate.
- [X] The issue still exists against the latest stable version of Elementor.

### Description

3 minute demo video: https://www.loom.com/share/7342035530544da59861e16a929b4824?sid=6171fd94-b1c8-4c6a-a602-d8defbac7809

If I copy and paste containers across Elementor sites (using the Elementor feature, not a 3rd party plugin), or if I export and import a template, my gap values do not transfer properly. I can paste as normal within the same site, but I cannot paste across without the gaps fouling. Some of my gaps are not "linked", meaning they have different column and row gap values.

If I inspect the JSON from an export file, the "size" value is null. I'm not sure if this is a problem:

```json
"id": "b871f7",
          "settings": {
            "flex_gap": {
              "column": "var(--kg_gap-600p)",
              "row": "var(--kg_gap-400p)",
              "isLinked": false,
              "unit": "custom",
              "size": null
            },
            "flex_gap_tablet": {
              "column": "var(--kg_gap-400p)",
              "row": "var(--kg_gap-300p)",
              "isLinked": false,
              "unit": "custom",
              "size": null
            },
            "flex_gap_mobile": {
              "column": "var(--kg_gap-200p)",
              "row": "var(--kg_gap-150p)",
              "isLinked": false,
              "unit": "custom",
              "size": null
            },

```

When I import or cross-site paste my containers, my gap values are inconsistent or change to a '10' and not the right value from the pasted site. Some containers will have the right gaps, but only for some breakpoints — not all.

It's a lightweight install and the theme is Hello with no child theme installed for both sites.

Under Elementor Features settings, "Additional Custom Breakpoints" is inactive.

### Steps to reproduce

1) Create a container and manually set the gap values. Make them different for desktop, tablet, and mobile. We used variables for our gap values and place the PX values in our CSS.

2) Put some nested containers in the parent container and set some manual gap values. Make them different for desktop, tablet, and mobile.

3) Make sure you have a global gap value set under Site Settings > Layout. We use a variable instead of a PX value in ours and set the PX value via CSS.

4) Copy the parent container and attempt to paste into a new site instance using "Paste from other site" OR export the original container(s) in a template and import the template into a new site

### Isolating the problem

- [X] This bug happens with only Elementor plugin active (and Elementor Pro).
- [X] This bug happens with a Blank WordPress theme active ([Hello theme](https://wordpress.org/themes/hello-elementor/)).
- [X] I can reproduce this bug consistently following the steps above.

### System Info
<details>
<summary>Click to reveal</summary>

````txt


					
== Server Environment ==
	Operating System: Linux
	Software: nginx/1.25.2
	MySQL version: MariaDB Server v10.4.25
	PHP Version: 8.1.24
	PHP Memory Limit: 256M
	PHP Max Input Vars: 10000
	PHP Max Post Size: 100M
	GD Installed: Yes
	ZIP Installed: Yes
	Write Permissions: All right
	Elementor Library: Connected

== WordPress Environment ==
	Version: 6.3.2
	Site URL: https://mysite.com
	Home URL: https://mysite.com
	WP Multisite: No
	Max Upload Size: 100 MB
	Memory limit: 40M
	Max Memory limit: 256M
	Permalink Structure: /%postname%/
	Language: en-US
	Timezone: America/Denver
	Debug Mode: Inactive

== Theme ==
	Name: Hello Elementor
	Version: 2.9.0
	Author: Elementor Team
	Child Theme: No

== User ==
	Role: administrator
	WP Profile lang: en_US
	User Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/16.6 Safari/605.1.15

== Active Plugins ==
	Code Snippets Pro
		Version: 3.5.1
		Author: Code Snippets Pro

	Elementor
		Version: 3.17.1
		Author: Elementor.com

	Elementor Pro
		Version: 3.16.2
		Author: Elementor.com

	Image optimization service by Optimole
		Version: 3.10.0
		Author: Optimole


== Must-Use Plugins ==
	Pantheon MU Plugin Loader
		Version: 1.0
		Author: Pantheon Systems


== Features ==
	Custom Fonts: 0
	Custom Icons: 0

== Integrations ==
	


== Elementor Experiments ==
	Optimized DOM Output: Active
	Improved Asset Loading: Active
	Improved CSS Loading: Active
	Inline Font Icons: Active
	Additional Custom Breakpoints: Inactive
	admin_menu_rearrangement: Inactive by default
	Flexbox Container: Active
	Upgrade Swiper Library: Active
	Grid Container: Inactive
	Default to New Theme Builder: Active by default
	Hello Theme Header & Footer: Inactive
	Editor Top Bar: Inactive by default
	Optimized Gutenberg Loading: Inactive by default
	Build with AI: Inactive by default
	Landing Pages: Inactive
	Nested Elements: Active
	Lazy Load Background Images: Inactive
	Optimize Image Loading: Inactive by default
	Global Style Guide: Inactive
	Page Transitions: Active by default
	Notes: Active by default
	Form Submissions: Active by default
	Scroll Snap: Inactive
	Menu: Inactive
	Taxonomy Filter: Inactive


== Log ==
	
PHP: showing 4 of 4PHP: 2023-09-28 19:54:57 [notice X 1][../wp-content/plugins/elementor/core/upgrade/manager.php::70] version_compare(): Passing null to parameter #1 ($version1) of type string is deprecated [array (
  'trace' => '
#0: Elementor\Core\Logger\Manager -> shutdown()
',
)]
PHP: 2023-09-28 20:00:39 [warning X 2][../wp-content/plugins/elementor/core/common/modules/ajax/module.php::175] Undefined array key "data" [array (
  'trace' => '
#0: Elementor\Core\Logger\Manager -> shutdown()
',
)]
PHP: 2023-09-28 14:23:11 [error X 2][../wp-content/plugins/elementor-pro/modules/theme-builder/classes/conditions-manager.php::304] Uncaught Error: Call to a member function update_meta() on null in ../wp-content/plugins/elementor-pro/modules/theme-builder/classes/conditions-manager.php:304
Stack trace:
#0 ../wp-content/plugins/elementor-pro/modules/theme-builder/classes/conditions-manager.php(177): ElementorPro\Modules\ThemeBuilder\Classes\Conditions_Manager->save_conditions(11, Array)
#1 [internal function]: ElementorPro\Modules\ThemeBuilder\Classes\Conditions_Manager->ajax_save_theme_template_conditions(Array, Object(Elementor\Core\Common\Modules\Ajax\Module))
#2 ../wp-content/plugins/elementor/core/common/modules/ajax/module.php(175): call_user_func(Array, Array, Object(Elementor\Core\Common\Modules\Ajax\Module))
#3 ../wp-includes/class-wp-hook.php(310): Elementor\Core\Common\Modules\Ajax\Module->handle_ajax_request('')
#4 ../wp-includes/class-wp-hook.php(334): WP_Hook->apply_filters('', Array)
#5 ../wp-includes/plugin.php(517): WP_Hook->do_action(Array)
#6 ../wp-admin/admin-ajax.php(188): do_action('wp_ajax_element...')
#7 {main}
  thrown [array (
  'trace' => '
#0: Elementor\Core\Logger\Manager -> shutdown()
',
)]
PHP: 2023-10-20 03:30:34 [warning X 1][../wp-content/plugins/elementor/includes/elements/container.php::1847] Array to string conversion [array (
  'trace' => '
#0: Elementor\Core\Logger\Manager -> shutdown()
',
)]

Log: showing 20 of 262023-10-09 09:56:56 [info] Elementor data updater process has been queued. [array (
  'plugin' => 'Elementor',
  'from' => '3.16.4',
  'to' => '3.16.5',
)]
2023-10-09 09:57:11 [info] elementor-pro::elementor_pro_updater Started 
2023-10-09 09:57:11 [info] Elementor Pro/Upgrades - _on_each_version Start  
2023-10-09 09:57:11 [info] Elementor Pro/Upgrades - _on_each_version Finished 
2023-10-09 09:57:11 [info] Elementor data updater process has been completed. [array (
  'plugin' => 'Elementor Pro',
  'from' => '3.14.1',
  'to' => '3.16.2',
)]
2023-10-09 09:57:11 [info] Elementor data updater process has been queued. [array (
  'plugin' => 'Elementor Pro',
  'from' => '3.14.1',
  'to' => '3.16.2',
)]
2023-10-18 11:32:30 [info] elementor::elementor_updater Started 
2023-10-18 11:32:30 [info] Elementor/Upgrades - _on_each_version Start  
2023-10-18 11:32:31 [info] Elementor/Upgrades - _on_each_version Finished 
2023-10-18 11:32:31 [info] Elementor/Upgrades - _v_3_17_0_site_settings_updates Start  
2023-10-18 11:32:31 [info] Elementor/Upgrades - _v_3_17_0_site_settings_updates Finished 
2023-10-18 11:32:31 [info] Elementor data updater process has been completed. [array (
  'plugin' => 'Elementor',
  'from' => '3.16.5',
  'to' => '3.16.6',
)]
2023-10-18 11:32:31 [info] Elementor data updater process has been queued. [array (
  'plugin' => 'Elementor',
  'from' => '3.16.5',
  'to' => '3.16.6',
)]
2023-10-25 17:56:43 [info] elementor::elementor_updater Started 
2023-10-25 17:56:43 [info] Elementor/Upgrades - _on_each_version Start  
2023-10-25 17:56:43 [info] Elementor/Upgrades - _on_each_version Finished 
2023-10-25 17:56:43 [info] Elementor/Upgrades - _v_3_17_0_site_settings_updates Start  
2023-10-25 17:56:43 [info] Elementor/Upgrades - _v_3_17_0_site_settings_updates Finished 
2023-10-25 17:56:43 [info] Elementor data updater process has been completed. [array (
  'plugin' => 'Elementor',
  'from' => '3.16.6',
  'to' => '3.17.1',
)]
2023-10-25 17:56:43 [info] Elementor data updater process has been queued. [array (
  'plugin' => 'Elementor',
  'from' => '3.16.6',
  'to' => '3.17.1',
)]

JS: showing 3 of 3JS: 2023-10-19 00:01:59 [error X 13][../wp-content/plugins/elementor/assets/js/frontend.min.js?ver=3.16.6:2:30893] elementor.settings.page is undefined 
JS: 2023-10-19 16:06:45 [error X 3][../wp-admin/load-scripts.php?c=1&loadchunk_0=jquery-core,jquery-migrate,utils,moxiejs,plupload&ver=6.3.2:2:28722] Cannot read properties of undefined (reading 'value') 
JS: 2023-10-26 02:03:52 [error X 3][../wp-admin/load-scripts.php?c=1&loadchunk_0=jquery-core,jquery-migrate,utils,moxiejs,plupload&ver=6.3.2:2:28729] undefined is not an object (evaluating 'l[0].value') 



== Elementor - Compatibility Tag ==
	
	Elementor Pro: Incompatible

== Elementor Pro - Compatibility Tag ==
	

				

````

</details>

**Comment by JuliaArr** [comment]
Hello! I experience the same problem with exporting the templates with containers with custom gaps, and all that I could google on the Internet is your bug report. Could you please let me know do you have any update from Elementor on the matter? 
Thanks a lot!

**Comment by KingGrizzly** [comment]
Hi Julia, I have no update. This seems very odd because I view this as a major bug. Export/import and cross site paste being broken is a big deal — it's really hurting our agency. We pre-built containers we can use as a starter for each build and I can't import them. We were planning to release these containers to the community, but can't do that either because no one could import them while this is an issue.

Are your gap values numbers or are you using a custom value like CSS or CSS variables?

Are your gap values the same for row and column?

Thanks, I'm really confused why almost no one is reporting this.
Brian

**Comment by KingGrizzly** [comment]
This is good info for Elementor. @nicholaszein , Julia was able to reproduce with a lightweight test case. I think this could mean no Elementor sites can properly export/import or cross site paste properly.

**Comment by KingGrizzly** [comment]
I'm mystified how @JuliaArr is the only other person reporting this issue and how it hasn't been addressed for over two months. We are losing hours on every build since we can't import or cross copy/paste.

**Comment by JuliaArr** [comment]
It's a huge problem for me too. I made some additional tests and can see that the problem is on the importing stage. 
I set a custom Row gap https://prnt.sc/tUyBkN2yjR7o, save it as a template and export, and the code in the template is okay: https://prnt.sc/43fyZjRvs1aK

`{"content":[{"id":"7ad40b18","settings":{"flex_gap":{"column":"20","row":"80","isLinked":false,"unit":"px","size":20}},"elements":[],"isInner":false,"elType":"container"}],"page_settings":[],"version":"0.4","title":"test1","type":"page"}`

But after importing the gap is always set to the default 20px https://prnt.sc/i8FuSohZAXDL

`{"column":"20","row":"20","isLinked":true,"unit":"px","size":20}`

The problem is still there on Elementor Version 3.18.3

**Comment by JuliaArr** [comment]
I also found out that the issue is exactly with the Gaps that are not Linked https://prnt.sc/rDrv7lRfBF4c
In that situation the code after export is:  `{"column":"30","row":"90","isLinked":false,"unit":"px","size":30}` - as you can see  "isLinked" is set to false, and there is also the "size" value which is set to 30 (as the Column Gap).
After numerous tests I can affirm that after importing the Gaps values are always linked and equal to this size value https://prnt.sc/qTyvGPi-quhb

So there is no possibility at the moment to transfer not linked Gaps.

**Comment by KingGrizzly** [comment]
@nicholaszein Please see Julia's findings. Export/import and cross site pasting are broken.

**Comment by lensbreak** [comment]
Lets see if we have to wait 1 years until this is fixed.

**Comment by KingGrizzly** [comment]
@lensbreak This was tagged just recently, so I'm hopeful the team can get to it soon as well as: https://github.com/elementor/elementor/issues/24178

**Comment by JuliaArr** [comment]
A new Elementor update was released 2 days ago, but it doesn't include a fix for this problem. Are there any updates on this issue?

**Comment by KingGrizzly** [comment]
I'm waiting for this to get addressed so I can release a project to the Elementor community. Can't give away the template file currently, because it won't import properly.

**Comment by KingGrizzly** [comment]
It's been since last October and importing + cross site pasting are still broken. I don't understand how a core feature like this can be overlooked for bug squashing. @shilo-ey Any chance someone can look into this? Gap values change on import and cross-site pasting. @JuliaArr pinpointed the issue: https://github.com/elementor/elementor/issues/24111#issuecomment-1878853229 Thanks.

**Comment by mpolek** [comment]
My solution is to change the two places in elementor/assets/js/admin.js and admin.min.js where it has:

if (!elementor.config.user.is_administrator) {

to:

if (!elementorAdmin.config.user.is_administrator) {

It's basically just a typo and easily fixable.

**Comment by KingGrizzly** [comment]
@mpolek Seems like an easy bug for Elementor to fix. Any idea why this would cause gap values to change on import or cross site paste?

**Comment by mpolek** [comment]
I believe the issue is that the typo causes an exception to be thrown in the JavaScript. Processing stops at that point. When I try to import a template, my experience is that I click the "Import Now" button and the process just stalls. The button is greyed out, but the import never completes. Errors can be different depending on what action is being taken.

I went ahead and opened a support ticket with the information for the fix. Perhaps that will result in a fix, as I suspect no one is looking here. :)

**Comment by KingGrizzly** [comment]
@mpolek Thanks. I hope it helps and gets addressed soon. Not 100% sure we have the same problem as my imports weren't timing out, just changing gap variables when they had differing row and column gaps.

**Comment by mpolek** [comment]
Well... If you have sample that exhibits the "bad" behavior with different gaps, try applying the fix and then importing it again. See if the gaps are fixed. Let me know what happens.

I wrote the bugfix on two threads that seemed to be related. If they are really not related, then I'll want to remove my comments from the gaps thread.

**Comment by KingGrizzly** [comment]
I demonstrate the issue exactly in this video which I will take down from YouTube once the problem is fixed. @nicholaszein 

https://youtu.be/qbTNnIpKXYc

**Comment by mpolek** [comment]
I believe they fixed it already. You'll notice that admin.js was updated in a recent release.

**Comment by KingGrizzly** [comment]
@mpolek I'm not so sure. This video was made on 3.20.1 Core and 3.20.0 Pro.

**Comment by karmaral** [comment]
Just dropping by to add a +1 on the issue. In my case, it's truncating the decimals out of the values (1.5em gets turned into 1em, 0.25 to 0 and so on).

It seems the pasting is reading the `size` value instead of the `column`/`row`, and that `size` doesn't appear to like decimals.

[`paste-area.js`](https://github.com/elementor/elementor/blob/482258ce1fe9ddf60883ee8c20e87ade146817ff/assets/dev/js/editor/document/elements/commands/paste-area.js#L42) seems to call the `'document/ui/paste'` command, which in theory it's the same called when doing local editor copy-pasting, and the values _are_ maintained properly when copy-pasting locally.

After [digging through some of the code](https://github.com/elementor/elementor/blob/61e8e41c60ec80517f547799f1a78ef35f09a2db/includes/controls/groups/flex-container.php#L168), I'm suspecting this `size` property is legacy and there might be an issue with the conversion. Though not entirely sure it's this, I can definitely see a `parseInt` [here](https://github.com/elementor/elementor/blob/61e8e41c60ec80517f547799f1a78ef35f09a2db/assets/dev/js/editor/controls/gaps.js#L39) that might be causing my particular issue as there are some mappings that are applied that convert size to column. Maybe it's all related.

Now I'd only need to try and make a local dev version work to test these changes... Too bad it's not as clearly document as the rest of the things.

**Comment by KingGrizzly** [comment]
@karmaral Did you have a chance to test your thoughts on a dev version? We are approaching 6 months with this core and advertised functionality not working. I tried to draw attention to it on Facebook out of sheer desperation (since the technical user forums were shut down) and was reprimanded for doing so. Disconcerting customer experience.

**Comment by embpdaniel** [comment]
I have a similar issue. I created a footer template that has 115px column gap / 20px row gap on desktop, and 115px column gap / 37px row gap on smaller screens. When I export the kit the values are correct on the template's .json file. But when I import the kit the value for both column and row gap on all sizes is 115px.

Unfortunate that this isn't working as intended and I will need to delay releasing my kit because of this.

**Comment by KingGrizzly** [comment]
@embpdaniel We've been trying to release something to the Elementor community since last year but have the same problem: it won't import properly. Same is true for copy layouts from one site to another or making a set of starter templates for our own builds. Gaps bonk.

**Comment by canaryninja** [comment]
+1

**Comment by KingGrizzly** [comment]
Cross site import/export and copy/paste are still broken. It has been 8 months.

**Comment by bentographics** [comment]
+1 Same issue here. Also with exporting and importing templates.

My only workaround was to copy over the elements, after that ues copy & "paste style" for each element individually. But it's a pain.

**Comment by KingGrizzly** [comment]
Check this out: A demonstration of the bug on two clean installs and some thoughts about what may be going on. I think it is tied to Elementor's migration from the single gap value to the double row/column gap values:

https://www.loom.com/share/adda9c50442b4135b2ab10dec759d42b?sid=852315b8-0483-4716-b76a-b99487d6110e

**Comment by KingGrizzly** [comment]
This problem with core copy/paste and export/import features remains unaddressed since October of 2023 despite acknowledgement by Elementor Support they can reproduce the issue. I don't understand how this is possible or acceptable from a leader in market share. If nothing else, we deserve communication around the issue. This is coming from a big Elementor fan and promoter. This needs to be fixed or at least no longer promoted as a feature of the product.

**Comment by captainplod** [comment]
Has anyone found a solution for this issue? It's crucial that it gets fixed!

**Comment by KingGrizzly** [comment]
Nope. It's really concerning how we can't get a fix or even proper engagement on this major bug. I can usually get bug fixes inside of a week from most other plugins once they realize the bug is real.

**Comment by bentographics** [comment]
Not a real solution and prone to manual errors, but we're handling it this way now:

1. Cross copy/paste sections, or inport templates (with said gap issue).
2. Manually Copy and "Paste Style" to elements which are affected. Cross-site "Paste Style" seems to preserve the correct gap values.

<img width="200" alt="Screen Shot 2024-09-02 at 11 55 19" src="https://github.com/user-attachments/assets/b0b08c55-f368-4b50-8cd4-92a4ac1f7c5d">


It's still less time consuming than manually setting up all affected elements again.
Hope this helps someone.

p.s. And yes, not acceptable how Elementor broke this crucial feature and fails to solve it in a timely manner.

**Comment by KingGrizzly** [comment]
@nicholaszein Thank you for the activity on this. Doubt you need any help, but I can send you two clean installs to experiment with if useful.

**Comment by KingGrizzly** [comment]
Going on 11 months of cross-site copy/paste and export/import being broken in Elementor.

**Comment by coulierp** [comment]
This issue has been a thorn in my side for a long time. It's frustrating having to fix the gap manually every time you cross-domain copy/paste a container with custom gap values.

I'd love to see this get fixed.

**Comment by KingGrizzly** [comment]
@HODAKdev Sorry, we did something similar: made a framework for the Elementor community but haven't been able to release it because all of the uneven gap settings in the kit will zero out on import. This has been the case for almost a year now. I can't fathom why Elementor hasn't fixed this and it's basically impossible to get any meaningful interaction about it. Copy/paste and import/export are core functionalities they advertise.

Can you explain or demonstrate your temporary fix? Not sure I fully understand.

**Comment by KingGrizzly** [comment]
@HODAKdev I understand how to paste styles within the same website, but how about across sites? If I copy or import to a new site, then uneven gaps go to '0'. If I try to 'Paste Style', it will only do so from the current site as opposed to the original site.

Plus, we'd be importing a lot of content so copying and pasting styles would be time consuming and easy to make mistakes with.

Unless I'm misunderstanding what you are explaining?

**Comment by KingGrizzly** [comment]
@HODAKdev No worries, you are just finding what I did: that there is a bug with gaps.

**Comment by AF1QUE** [comment]
@nicholaszein Any update on this? We had to redo thousands of templates that we previously created using the unlinked gap system as a workaround for this bug. This is a major issue as it breaks the container gap system when users copy containers between websites.

**Comment by KingGrizzly** [comment]
@AF1QUE I think @hein-obox may be working on this now. See his comment on this thread: https://github.com/elementor/elementor/issues/24178 . I feel your pain! We heavily rely on controlling both column and row gaps through the usage of variables in the gap fields.

**Comment by nicholaszein** [comment]
Hi @AF1QUE, we are working on this bug! 😉

**Comment by AF1QUE** [comment]
> Hi @AF1QUE, we are working on this bug! 😉

That's quite an update, thank you! I'd also like to bring another copy-paste issue to your attention, which prevents users from copying widgets properly containing SVG assets.

https://github.com/elementor/elementor/issues/21136

**Comment by nicholaszein** [comment]
Hi everyone!

We have great news! 🙌 

### 📢 We're happy to announce that this issue was resolved with the release of Elementor v3.26.0! 🥳 

✅ Feel free to check it out and update your plugin to the new version!

> _Check the changelog for more information:_
> [Changelog of Elementor and Elementor Pro](https://elementor.com/help/changelog-of-elementor/)

Cheers 🥂
